.doom.d/config.org
#+TITLE: Config
#+STARTUP: overview

Lots stolen from [[https://gitlab.com/thatonelutenist/dotfiles/-/blob/master/emacs.d/settings.org][Nathan's awesome config]].

* Identity
Some functionality uses this to identify you, e.g. GPG configuration, email
clients, file templates and snippets.

#+BEGIN_SRC elisp
(setq user-full-name "Jani Mustonen"
      user-mail-address "janijohannes@kapsi.fi")

(setq auth-sources '("~/.authinfo"))
#+END_SRC

* Styles
** Doom hints
Doom exposes five (optional) variables for controlling fonts in Doom. Here
are the three important ones:

+ `doom-font'
+ `doom-variable-pitch-font'
+ `doom-big-font' -- used for `doom-big-font-mode'; use this for
  presentations or streaming.

They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
font string. You generally only need these two:

#+BEGIN_SRC elisp :tangle no
;; Ignored code block for documentation
(setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "sans" :size 13))
#+END_SRC

There are two ways to load a theme. Both assume the theme is installed and
available. You can either set `doom-theme' or manually load a theme with the
`load-theme' function. This is the default:
** Config
#+BEGIN_SRC elisp :tangle packages.el
(package! rainbow-delimiters)
(package! color-identifiers-mode)
#+END_SRC

Best themes: `doom-vibrant`, `doom-moonlight`, `doom-gruvbox`, `doom-acario-dark`, `doom-horizon`
#+BEGIN_SRC elisp
(setq doom-theme 'doom-gruvbox)
(setq inhibit-x-resources t)

(use-package! rainbow-delimiters
  :hook (prog-mode . rainbow-delimiters-mode))
(use-package! color-identifiers-mode
  :config
  (add-hook 'after-init-hook #'global-color-identifiers-mode)
  (setq color-identifiers:min-color-saturation 0.3)
  (setq color-identifiers:color-luminance 0.7))
#+END_SRC

This determines the style of line numbers in effect. If set to `nil', line numbers are disabled. For relative line numbers, set this to `relative'.
#+BEGIN_SRC elisp
(setq display-line-numbers-type t)
#+END_SRC

** Fonts
#+BEGIN_SRC elisp
(setq doom-font (font-spec :family "Fira Code" :size 20)
      doom-variable-pitch-font (font-spec :family "Roboto" :size 21 :weight 'bold) ; inherits `doom-font''s :size
      doom-unicode-font (font-spec :family "Fira Code" :size 20)
      doom-big-font (font-spec :family "Fira Code" :size 28))
#+END_SRC

** Mixed pitch
Depends on [[doom-modules:ui/zen/config.el][doom's zen-mode's config]].
#+BEGIN_SRC elisp :tangle no
(use-package! mixed-pitch
  :config
  (setq mixed-pitch-set-height 't)
  (add-hook 'org-mode-hook #'mixed-pitch-mode)
  (add-hook 'markdown-mode-hook #'mixed-pitch-mode))
#+END_SRC

** Fira code ligatures
Adds support for code ligatures, requires [[https://github.com/tonsky/FiraCode/files/412440/FiraCode-Regular-Symbol.zip][Fira Code Symbol]] font for this to work.
#+BEGIN_SRC elisp :tangle packages.el
(package! fira-code-mode)
#+END_SRC

#+BEGIN_SRC elisp
(use-package! fira-code-mode
  :custom (fira-code-mode-disabled-ligatures '("x" "#["))
  :hook prog-mode)
#+END_SRC

* Org mode configuration
[[https://orgmode.org/manual/index.html][org-mode manual]]
[[https://orgmode.org/worg/org-contrib/][org-con trib index]]
** base

#+BEGIN_SRC elisp
(setq org-directory "~/org")

;; Restore default TAB behavior
(after! evil-org
  (remove-hook 'org-tab-first-hook #'+org-cycle-only-current-subtree-h))
#+END_SRC

The capturing target is the default for new templates, but it's not actually respected by the existing ones. Use `SPC X` to capture.

#+BEGIN_SRC elisp
(setq org-default-notes-file (concat org-directory "/index.org"))
#+END_SRC

** org-notify
Comes from org-contrib.
#+BEGIN_SRC elisp
(use-package! org-notify
  :config
  (org-notify-start)
  (org-notify-add 'default)
  (org-notify-add 'yes
                  '(:time "1h" :period "10m" :duration 20
                    :actions -notify/window)
                  '(:time "2h" :period "10m" :actions -message)
                  '(:time "1d" :duration 300 :actions -notify/window)))
#+END_SRC
** style
#+BEGIN_SRC elisp
(add-hook 'org-mode-hook #'visual-line-mode)
#+END_SRC
* Git
Note: magit uses `auth-sources` to authorize with Github & Gitlab.

#+BEGIN_SRC elisp
(setq magit-revision-show-gravatars '("^Author:     " . "^Commit:     "))
#+END_SRC

* Rust
Uses[[https://github.com/brotzeit/rustic][ Rustic]].

** Flycheck fix
Rustic's default clippy config doesn't work on stable rust yet. See [[https://github.com/brotzeit/rustic#flycheck]]

#+BEGIN_SRC elisp
;; patch rustic for stable rust
(after! rustic
  (setq rustic-flycheck-clippy-params "--message-format=json"))
#+END_SRC
** Use clippy as checker on LSP mode
#+BEGIN_SRC elisp
(setq lsp-rust-analyzer-cargo-watch-command "clippy")
#+END_SRC
* Markdown
#+BEGIN_SRC elisp
(add-hook 'markdown-mode-hook #'visual-line-mode)
#+END_SRC
* Projectile
This is a bit terrible because it depends on load order..
#+BEGIN_SRC elisp
(after! projectile
  (setq projectile-project-root-files '(".git")))
#+END_SRC
* Calendar
** Week starts on a Monday.
#+BEGIN_SRC elisp
(after! calendar
  (setq calendar-week-start-day 1))
#+END_SRC
* YaSnippet
** Keybindings
#+BEGIN_SRC elisp
(after! yasnippet
  (define-key yas-minor-mode-map (kbd "C-§") #'yas-expand))
#+END_SRC
* Javascript & friends
** Lint
#+BEGIN_SRC elisp
(setq lsp-eslint-server-command
   '("node"
     "/home/jani/.vscode/extensions/dbaeumer.vscode-eslint-2.1.8/server/out/eslintServer.js"
     "--stdio"))
#+END_SRC
* Ediff
** Combine regions
Bind `d` to combine both regions
#+begin_src elisp
(defun ediff-copy-both-to-C ()
  (interactive)
  (ediff-copy-diff ediff-current-difference nil 'C nil
                   (concat
                    (ediff-get-region-contents ediff-current-difference 'A ediff-control-buffer)
                    (ediff-get-region-contents ediff-current-difference 'B ediff-control-buffer))))
(defun add-d-to-ediff-mode-map () (define-key ediff-mode-map "d" 'ediff-copy-both-to-C))
(add-hook 'ediff-keymap-setup-hook 'add-d-to-ediff-mode-map)
#+end_src
* Desktop
#+begin_src elisp
(add-hook 'after-init-hook #'desktop-read)
#+end_src
* TabNine
#+begin_src elisp :tangle packages.el
(package! company-tabnine)
#+end_src

#+begin_src elisp
(use-package! company-tabnine)
(after! company-tabnine
  (set-company-backend! 'python-mode #'company-tabnine)
  (set-company-backend! 'rustic-mode #'company-tabnine)
  ;; Trigger completion immediately.
  (setq company-idle-delay 0)

  ;; Number the candidates (use M-1, M-2 etc to select completions).
  (setq company-show-numbers t))
#+end_src
* Doom hints
Here are some additional functions/macros that could help you configure Doom:

- `load!' for loading external *.el files relative to this one
- `use-package!' for configuring packages
- `after!' for running code after a package has loaded
- `add-load-path!' for adding directories to the `load-path', relative to
  this file. Emacs searches the `load-path' when you load packages with
  `require' or `use-package'.
- `map!' for binding new keys

To get information about any of these functions/macros, move the cursor over
the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
This will open documentation for it, including demos of how they are used.

You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
they are implemented.


.doom.d/config.el
(setq user-full-name "Jani Mustonen"
      user-mail-address "janijohannes@kapsi.fi")

(setq auth-sources '("~/.authinfo"))

(setq doom-theme 'doom-gruvbox)
(setq inhibit-x-resources t)

(use-package! rainbow-delimiters
  :hook (prog-mode . rainbow-delimiters-mode))
(use-package! color-identifiers-mode
  :config
  (add-hook 'after-init-hook #'global-color-identifiers-mode)
  (setq color-identifiers:min-color-saturation 0.3)
  (setq color-identifiers:color-luminance 0.7))

(setq display-line-numbers-type t)

(setq doom-font (font-spec :family "Fira Code" :size 20)
      doom-variable-pitch-font (font-spec :family "Roboto" :size 21 :weight 'bold) ; inherits `doom-font''s :size
      doom-unicode-font (font-spec :family "Fira Code" :size 20)
      doom-big-font (font-spec :family "Fira Code" :size 28))

(use-package! fira-code-mode
  :custom (fira-code-mode-disabled-ligatures '("x" "#["))
  :hook prog-mode)

(setq org-directory "~/org")

;; Restore default TAB behavior
(after! evil-org
  (remove-hook 'org-tab-first-hook #'+org-cycle-only-current-subtree-h))

(setq org-default-notes-file (concat org-directory "/index.org"))

(use-package! org-notify
  :config
  (org-notify-start)
  (org-notify-add 'default)
  (org-notify-add 'yes
                  '(:time "1h" :period "10m" :duration 20
                    :actions -notify/window)
                  '(:time "2h" :period "10m" :actions -message)
                  '(:time "1d" :duration 300 :actions -notify/window)))

(add-hook 'org-mode-hook #'visual-line-mode)

(setq magit-revision-show-gravatars '("^Author:     " . "^Commit:     "))

;; patch rustic for stable rust
(after! rustic
  (setq rustic-flycheck-clippy-params "--message-format=json"))

(setq lsp-rust-analyzer-cargo-watch-command "clippy")

(add-hook 'markdown-mode-hook #'visual-line-mode)

(after! projectile
  (setq projectile-project-root-files '(".git")))

(after! calendar
  (setq calendar-week-start-day 1))

(after! yasnippet
  (define-key yas-minor-mode-map (kbd "C-§") #'yas-expand))

(setq lsp-eslint-server-command
   '("node"
     "/home/jani/.vscode/extensions/dbaeumer.vscode-eslint-2.1.8/server/out/eslintServer.js"
     "--stdio"))

(defun ediff-copy-both-to-C ()
  (interactive)
  (ediff-copy-diff ediff-current-difference nil 'C nil
                   (concat
                    (ediff-get-region-contents ediff-current-difference 'A ediff-control-buffer)
                    (ediff-get-region-contents ediff-current-difference 'B ediff-control-buffer))))
(defun add-d-to-ediff-mode-map () (define-key ediff-mode-map "d" 'ediff-copy-both-to-C))
(add-hook 'ediff-keymap-setup-hook 'add-d-to-ediff-mode-map)

(add-hook 'after-init-hook #'desktop-read)

(use-package! company-tabnine)
(after! company-tabnine
  (set-company-backend! 'python-mode #'company-tabnine)
  (set-company-backend! 'rustic-mode #'company-tabnine)
  ;; Trigger completion immediately.
  (setq company-idle-delay 0)

  ;; Number the candidates (use M-1, M-2 etc to select completions).
  (setq company-show-numbers t))
